README.md
# Dinos x Robots Simulator

This project provides an API for simulating robots walking on a grid and fighting some dinosaurs - even though these poor creatures are already extinct. 

## Requirements

* [Java JDK 8](http://openjdk.java.net/install)
* [Leiningen 2.8.1](https://leiningen.org)

## Running

Run with command: `lein run`

The API will then be listening on port 8080.

## Features

* Create an empty grid (default 50 x 50)
* Place a robot in a certain position (robots by default are facing to the right)
* Place a dinosaur in a certain position (dinosaurs don't move)
* Issue commands to a robot, such as:
	- turn left or right
	- move forward or backwards
	- attack
* Display current board state
* Reset board

### Routes

* **GET /show-state**
	- Shows board current state.
	- Example: `curl http://localhost:8080/show-state`

* **POST /place-dino**
	- Places a dinosaur on a given position (passed on body of request)
	- Example: `curl --data "{\"x\":45,\"y\":0}" http://localhost:8080/place-dino`

* **POST /place-robot**
	- Places a robot in a given position (passed on body of request)
	- Example: `curl --data "{\"x\":45,\"y\":0}" http://localhost:8080/place-dino`

* **POST /robot-cmd/:op**
	- Execute a robot command
	- Valids operations (:op) are: `fwd-move` `rev-move` `rotate-left` `rotate-right` `attack`
	- Example: `curl --data "{\"x\":45,\"y\":0}" http://localhost:8080/robot-cmd/attack`

* **POST /reset-board**
	- Resets the board state to an empty grid
	- Example: `curl --data "{}" http://localhost:8080/reset-board`

## Illustration

This is how a **GET /show-state** would print the grid on a browser:

```
#####
#####
#####
#####
#####
```

Now, if we add a dinosaur on the [1 1] position: `curl --data "{\"x\":1,\"y\":1}" http://localhost:8080/place-dino`

This is how the board will look like:

```
#####
#D###
#####
#####
#####
```

Please notice this board is 5x5 for illustration purpose. The real board will be 50x50.

## Testing

Run all unit tests with command: `lein test`

## Documentation

An automatically generated documentation is available on the `doc` folder.

## Final Observations

* Issuing an invalid command will either return a 400 code or just ignore the command
* Attempting to place a robot (or dinosaur) into an occupied cell will replace it
* `/show-state` will refresh automatically after 2 seconds

## License

Copyright © 2018

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-08-10
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-08-10
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/dinos/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/dinos/compare/0.1.0...0.1.1


test/dinos/core_test.clj
(ns dinos.core-test
  (:require [clojure.test :refer :all]
            [dinos.core :refer :all]))

;; testing board functions
(deftest zeros-test
  (testing "should return lazy sequence of zeros"
    (is (instance? clojure.lang.LazySeq zeros))
    (is (= (take 1 zeros) '(0)))
    (is (= (take 3 zeros) '(0 0 0)))))

(deftest new-board-test
  (testing "creation of new board should return an NxN matrix with only zeros"
    (is (= (new-board 2) '((0 0), (0 0))))
    (is (not= (new-board 2) '((1 1), (1 1))))
    (is (=  (new-board 3) '((0 0 0), (0 0 0), (0 0 0))))))

(deftest place-in-row-test
  (testing "placement of an element on certain position of a list"
    (is (= (place-in-row [1 2] 1 "abc") [1 "abc"]))
    (is (= (place-in-row [1 2 3] 1 0) [1 0 3])))
  (testing "trying to place an element out of range"
    (is (thrown? Exception (place-in-row [1 2] "abc" 4)))))

(deftest place-in-board-test
  (testing "placement of an element of certain position of a grid/board"
    (is (= (place-in-board [[1]] [0 0] "testing") [["testing"]]))
    (is (= (place-in-board [[0 0] [0 0]] [0 0] 1) [[1 0] [0 0]])))
  (testing "trying to place an element out of range"
    (is (thrown? Exception (place-in-board [[1]] [1 2] "testing")))))

(deftest place-dino-test
  (testing "placement of dino into board"
    (is (= (place-dino [[0 0] [0 0]] [0 0]) [[2 0] [0 0]]))))

(deftest place-robot-test
  (testing "placement of robot into board"
    (is (not= (place-robot [[0]] [0 0]) [[[-1 0]]]))
    (is (= (place-robot [[0]] [0 0]) [[[1 0]]]))))

(deftest erase-position-test
 (testing "erasement of a position of the board"
  (is (= (erase-position [[1 0] [0 0]] [0 0]) [[0 0] [0 0]])))) 

(deftest get-item-test
  (testing "if function returns the element of given position of the board"
    (is (= (get-item [[-1]] [0 0]) -1))
    (is (not= (get-item [[1 1] [2 2]] [1 0]) 2))))

(deftest robot?-test
  (testing "if there's a robot on the given position"
    (let [a-board (new-board 10)]
      (is (robot? (place-robot a-board [0 1]) [0 1]))
      (is (not (robot? (place-robot a-board [0 1]) [1 1]))))))

(deftest dino?-test
  (testing "if there's a dino on the given position"
    (let [a-board (new-board 10)]
      (is (dino? (place-dino a-board [0 1]) [0 1]))
      (is (not (dino? (place-robot a-board [0 1]) [0 1]))))))

(deftest free?-test
  (testing "if given position of the board is a zero"
    (let [a-board (new-board 10)]
      (is (free? a-board [0 1]))
      (is (not (free? (place-robot a-board [0 1]) [0 1]))))))

(deftest valid?-test
  (testing "if a position is inside the board"
    (let [a-board (new-board 10)]
      (is (valid? a-board [0 0]))
      (is (valid? a-board [5 5]))
      (is (not (valid? a-board [10 11]))))))

;; testing robots commands
(deftest move-test
  (testing "if function returns the sum of given coordinates with a direction vector"
    (is (= (move [2 2] [-1 0]) [1 2]))
    (is (= (move [0 0] [1 0]) [1 0]))
    (is (= (move [2 2] [0 1]) [2 3] ))))

(deftest can-move?-test
  (let [a-board (new-board 10)]
    (testing "if move is still inside the board"
      (is (can-move? a-board [2 2]))
      (is (can-move? a-board [9 9])))
    (testing "if move is outside the range of the board"
      (is (not (can-move? a-board [10 10])))
      (is (not (can-move? a-board [-1 2]))))))

(deftest do-move-test
  (let [a-board (place-robot (new-board 10) [0 0])
        a-dino-board (place-dino a-board [1 0])]
    (testing "if robot by default moves 'to the right'"
      (is (= (do-move a-board [0 0] identity) (place-robot (new-board 10) [1 0]))))
    (testing "if robot can't visit a place with a dino on it"
      (is (= (do-move a-dino-board [0 0] identity) a-dino-board)))))

(deftest fwd-move-test
  (let [a-board (place-robot (new-board 10) [2 2])]
    (testing "if default robot walks for its 'right' on a fwd-move"
      (is (= (fwd-move a-board [2 2]) (place-robot (new-board 10) [3 2]))))))

(deftest rev-move-test
  (let [a-board (place-robot (new-board 10) [2 2])]
    (testing "if default robot walks for its 'left' on a rev-move"
      (is (= (rev-move a-board [2 2]) (place-robot (new-board 10) [1 2]))))))

(deftest turn-right-test
  (testing "if direction vector turns to the right"
    (is (= (turn-right [1 0]) [0 1]))
    (is (= (turn-right [0 1]) [-1 0]))
    (is (= (turn-right [-1 0]) [0 -1]))
    (is (= (turn-right [0 -1]) [1 0]))))

(deftest turn-left-test
  (testing "if direction vector turns to the left"
    (is (= (turn-left [1 0]) [0 -1]))
    (is (= (turn-left [0 -1]) [-1 0]))
    (is (= (turn-left [-1 0]) [0 1]))
    (is (= (turn-left [0 1]) [1 0]))))

(deftest rotate-right-test
  (testing "if robots change it's direction vector for a right rotation"
    (let [a-board (place-robot (new-board 10) [2 2])]
      (is (= (get-item (rotate-right a-board [2 2]) [2 2]) [0 1])))))

(deftest rotate-left-test
  (testing "if robots change it's direction vector for a left rotation"
    (let [a-board (place-robot (new-board 10) [2 2])]
      (is (= (get-item (rotate-left a-board [2 2]) [2 2]) [0 -1])))))

(deftest do-attack-test
  (testing "if dinos are erased from given position"
    (let [a-board (place-dino (new-board 10) [1 2])]
      (is (= (new-board 10) (do-attack a-board [1 2]))))))

(deftest attack-test
  (testing "if robot erases all dinos from (only) its adjacent cells"
    (let [a-board (place-robot (new-board 10) [2 2])
          a-dino-board (place-dino a-board [1 2])
          an-away-dino-board (place-dino a-board [5 5])]
      (is (= (attack an-away-dino-board [2 2]) an-away-dino-board))
      (is (= (attack a-dino-board [2 2]) a-board)))))

;; board state tests
(deftest get-state-test
  (testing "if initially the board state references a dimxdim grid with all zeros"
    (is (= (get-state) (new-board board-dim)))))

(deftest change-state-test
  (testing "if state changes with the application of a function"
    (is (= (change-state place-robot [1 1]) (place-robot (new-board board-dim) [1 1])))
    (is (= (get-state) (place-robot (new-board board-dim) [1 1])))))

(deftest reset-board-state-test
  (testing "if resetting the board state returns a new empty dimxdim grid"
    (is (= (reset-board-state) (new-board board-dim)))))

;; api logic tests
(deftest show-board-state-test
  (testing "if show-board-state returns a 200 code"
    (is (= 200 (:status (show-board-state)))))
  (testing "if there's a body"
    (is (not (nil? (:body (show-board-state)))))))

(deftest do-action-test
  (testing "if do-action returns a 400 for a bad request"
    (is (= 400 (:status (do-action {:body 1} place-robot))))))



src/dinos/core.clj
(ns dinos.core
  (:require [clojure.data.json :as json])
  (:require [org.httpkit.server :refer [run-server]]
            [compojure.core :refer :all]
            [compojure.route :as route])
  (:require [clojure.tools.logging :as log])
  (:gen-class))

(def free 0)
(def robot [1 0]) ;robots default direction is "to the right"
(def dino 2)

;;;;
;; Board operations
;;;;
(defn zeros*
  "generates a lazy sequence with only zeros"
  []
  (map (constantly 0) (range)))

(def zeros (zeros*))

(defn new-board
  "creates an empty board with dim x dim dimensions"
  ([dim cur]
   (if (= cur 0) '()
     (cons (take dim zeros) (new-board dim (dec cur)))))
  ([dim]
   (new-board dim dim)))

(defn place-in-row
  "receives a list and returns a new list with a modified element"
  [l pos value]
  (let [newl (take pos l)
        newl (concat newl (list value))
        newl (concat newl (drop (inc pos) l))]
    newl))

(defn place-in-board
  "receives a board and returns a new board with a modified element"
  [board [x y] value]
  (place-in-row board y (place-in-row (nth board y) x value)))

(defn place-dino
  "places a dino on the given position of the board"
  [board [x y]]
  (place-in-board board [x y] dino))

(defn place-robot
  "places a robot on the given position of the board"
  [board [x y]]
  (place-in-board board [x y] robot))

(defn erase-position
  "frees the given position of the board"
  [board [x y]]
  (place-in-board board [x y] free))

(defn get-item 
  "returns item on given position"
  [board [x y]]
  (nth (nth board y) x))

(defn robot?
  "verifies if there's a robot on given position"
  [board [x y]]
  (sequential? (get-item board [x y])))

(defn dino?
  "verifies if there's a dino on given position"
  [board [x y]]
  (= (get-item board [x y]) dino))

(defn free?
  "verifies if given position is free"
  [board [x y]]
  (= (get-item board [x y]) free))

(defn valid?
  "returns true if (x,y) is inside the board"
  [board [x y]]
  (let [dim (count board)]
    (and (>= x 0) (< x dim) (>= y 0) (< y dim))))

(defn prettify-row
  "converts a board row to a string with pretty html symbols"
  [row]
  (str (apply str (map #(if (= % 0) "#" 
                     (if (= % 2) "D" "R")) row)) "<br>"))

(defn pretty-board
  "prettify board for better representation"
  [board]
  (apply str (map prettify-row board)))

;;;;
;; robots operations
;;;;
(def up-vec [0 -1])
(def down-vec [0 1])
(def right-vec [1 0])
(def left-vec [-1 0])

;; movements
(defn move
  "returns new position based on given position and given direction"
  [[x y] dir]
  (map + [x y] dir))

(defn can-move?
  "returns true if robot can move to the given position"
  [board [x y]]
  (and (valid? board [x y]) (free? board [x y])))

(defn do-move
  "returns a new board with robot's position indicated by f.
  returns the board if move is not valid"
  [board [x y] f]
  (if (robot? board [x y])
    (let [item (get-item board [x y])
          dir (map f item)
          pos (move [x y] dir)]
      (if (can-move? board pos)
        (place-in-board (erase-position board [x y]) pos item)
      board))
    board))

(defn fwd-move
  "moves robot forward in according to its current direction"
  [board [x y]]
  (do-move board [x y] identity))
  
(defn rev-move
  "moves robot backwards in according to its current direction"
  [board [x y]]
  (do-move board [x y] -))

;; rotations
(defn turn-right
  [dir]
  (cond
    (= dir up-vec) right-vec
    (= dir right-vec) down-vec
    (= dir down-vec) left-vec
    (= dir left-vec) up-vec))

(defn turn-left
  [dir]
  (into [] (map - (turn-right dir))))

(defn rotate
  "changes robot direction clockwise or counterclockwise
  (determined by f)"
  [board [x y] f]
  (if (robot? board [x y])
    (let [pos (get-item board [x y])]
      (place-in-board board [x y] (f pos)))
    board))

(defn rotate-right
  [board [x y]]
  (rotate board [x y] turn-right))

(defn rotate-left
  [board [x y]]
  (rotate board [x y] turn-left))

;; attack
(defn do-attack
  "erases given position or returns the board if attack is not valid"
  [board [x y]]
  (if (and (valid? board [x y]) (dino? board [x y]))
    (erase-position board [x y])
    board))

(defn attack
  "makes robot from given position attack its adjacent positions"
  [board [x y]]
  (if (robot? board [x y])
    (let [left (move [x y] left-vec)
          right (move [x y] right-vec)
          down (move [x y] down-vec)
          up (move [x y] up-vec)]
      (reduce do-attack board [left right down up]))
    board))

;;;;
;; board state management
;;;;
(def board-dim 50)

(def board-state (atom (new-board board-dim)))

(defn get-state
  "returns the current state of the board"
  [& args]
  @board-state)

(defn change-state
  "receives a board function and change state to the return of this function"
  [f args]
  (swap! board-state f args))

(defn reset-board-state
  "resets board state to default value"
  [& args] ;; TODO: use swap! instead?
  (reset! board-state (new-board board-dim)))

;;;;
;; API logic
;;;;
(defn show-board-state
  []
  (let [response {:status 200
                  :headers {"Content-Type" "text/html"}
                  :body (str "<meta http-equiv=\"refresh\" content=\"2\"><body>" (str (pretty-board (get-state))) "</body>" )}]
    response))

(defn do-action
  [req f]
  (try
    (let [body (json/read-str (slurp (:body req)) :key-fn keyword)
          response {:status 200}]
      (change-state f [(:x body) (:y body)])
      response)
  (catch Exception ex
    ;(log/error "Error on " (:uri req) (.getMessage ex))
    {:status 400
     :body "bad request"})))

(defn place-dino-handler
  [req]
  (do-action req place-dino))

(defn place-robot-handler
  [req]
  (do-action req place-robot))

(defn robot-cmd-handler
  [req]
  (let [cmd (-> req :params :op)]
    (case cmd
      "fwd-move" (do-action req fwd-move)
      "rev-move" (do-action req rev-move)
      "attack" (do-action req attack)
      "rotate-left" (do-action req rotate-left)
      "rotate-right" (do-action req rotate-right)
      {:status 400
       :body "bad request"})))

(defn reset-board-handler
  [& _]
  (let [response {:status 200}]
    (reset-board-state)
    response))

(defroutes app
  (GET "/" [] "<h1>Welcome</h1>")
  (GET "/show-state" [] (show-board-state))
  (POST "/place-dino" [] place-dino-handler)
  (POST "/place-robot" [] place-robot-handler)
  (POST "/robot-cmd/:op" [] robot-cmd-handler)
  (POST "/reset-board" [] reset-board-handler)
  (route/not-found "<h1>Page not found</h1>"))


(defn -main [& args]
    (run-server app {:port 8080})
      (println "Server started on port 8080"))

